elks/fs/romfs/romfs-low.S

```diff
@@ -0,0 +1,250 @@
+; ROMFS - A tiny filesystem in ROM
+
+; This module implements low-level functions
+; to directly access the filesystem in ROM,
+; and used by the ROMFS high-level module.
+
+#define __ASSEMBLY__
+#include <linuxmt/config.h>
+
+	.text
+
+; Get superblock
+; int romfs_super_get (byte_t *);
+; TODO: add argument with size of destination buffer
+
+_romfs_super_get:
+
+	push  bp
+	mov   bp,sp
+	push  ds
+	push  di
+	push  si
+	push  cx
+
+	mov   ax,#CONFIG_ROMFS_BASE
+	mov   ds,ax
+
+	; TODO: check super block magic
+
+	xor   si,si
+	mov   cx,[si+6]  ; size of superblock
+	mov   di,[bp+4]  ; arg0
+
+	cld
+	rep
+	movsb
+
+	xor   ax,ax
+
+	pop   cx
+	pop   si
+	pop   di
+	pop   ds
+	pop   bp
+	ret
+
+
+; Get inode from index
+; int romfs_inode_get (word_t index, byte_t *);
+; TODO: add argument with size of destination buffer
+
+_romfs_inode_get:
```

Comment:
Should there be checks in these assembly routines, e.g. a bad value used for indirection doesn't cause us to go outside of the boundaries of the ROMFS area?

Reply:
If you evr want to use the ia16-elks-gcc compiler, you have to keep in mind the differences in usage if segment register ES. In this compiler, ES is used as scratchpad and can have anything, and its contents must be preserved across function call. This is, you should have to save ES, load it with the kernel data segment and restore its value before return.

Reply:
Sure. Indeed I won't rework these routines at all, as @lithoxs highlighted the point that these are actually not needed and will be replaced by more generic code in a next step.

Reply:
The struct inode_operations was modified in last commit. The bmap field is used, with another name. You can leave it with NULL, but the space has to be used. See files in fs/minix.